### ISSUE TITLE
REG: inconsistency in masked arrays with boolean array addressing

### ISSUE THREAD

**Issue by neutrinoceros** [numpy/numpy#19721]
Initial discovery: https://github.com/yt-project/yt/issues/3472
Later reported and discussed as https://github.com/matplotlib/matplotlib/issues/20843
Many thanks to @jklymak for contributing a pure-numpy test in comments. This report was edited to feature a slightly modified version of his test instead of the original example which included matplotlib code and was much harder to decipher.

### Reproducing code example:

```python
import numpy as np

a = np.ma.masked_array([0, 1], mask=[False, False])
b = np.ma.masked_array([0, 1], mask=[True, True])
mask = a < 1
b[mask] = a[mask]
expected_mask = [False, True]
np.testing.assert_array_equal(b.mask, expected_mask)
```

### Error message:
<details>
<summary>Traceback</summary>

```
╭─────────────────────────────── Traceback (most recent call last) ────────────────────────────────╮
│ /Users/robcleme/dev/yt-project/yt-py310/numpy/t.py:8 in <module>                                 │
│                                                                                                  │
│   5 mask = a < 1                                                                                 │
│   6 b[mask] = a[mask]                                                                            │
│   7 expected_mask = [False, True]                                                                │
│ ❱ 8 np.testing.assert_array_equal(b.mask, expected_mask)                                         │
│   9                                                                                              │
│                                                                                                  │
│ /Users/robcleme/dev/yt-project/yt-py310/numpy/numpy/testing/_private/utils.py:934 in             │
│ assert_array_equal                                                                               │
│                                                                                                  │
│    931 │                                                                                         │
│    932 │   """                                                                                   │
│    933 │   __tracebackhide__ = True  # Hide traceback for py.test                                │
│ ❱  934 │   assert_array_compare(operator.__eq__, x, y, err_msg=err_msg,                          │
│    935 │   │   │   │   │   │    verbose=verbose, header='Arrays are not equal')                  │
│    936                                                                                           │
│    937                                                                                           │
│                                                                                                  │
│ /Users/robcleme/dev/yt-project/yt-py310/numpy/numpy/testing/_private/utils.py:844 in             │
│ assert_array_compare                                                                             │
│                                                                                                  │
│    841 │   │   │   msg = build_err_msg([ox, oy], err_msg,                                        │
│    842 │   │   │   │   │   │   │   │   verbose=verbose, header=header,                           │
│    843 │   │   │   │   │   │   │   │   names=('x', 'y'), precision=precision)                    │
│ ❱  844 │   │   │   raise AssertionError(msg)                                                     │
│    845 │   except ValueError:                                                                    │
│    846 │   │   import traceback                                                                  │
│    847 │   │   efmt = traceback.format_exc()                                                     │
╰──────────────────────────────────────────────────────────────────────────────────────────────────╯
AssertionError:
Arrays are not equal

Mismatched elements: 1 / 2 (50%)
 x: array([ True,  True])
 y: array([False,  True])
```
</details>

### NumPy/Python version information:

The error is raised with the current dev versions of numpy and bisects down to 2504230 (#19244)
I used Python 3.10b2

**Comment by jklymak** [comment]
A simpler test that doesn't involve matplotlib I think is:

```python
a = np.ma.masked_array([0.0, 1.1], mask=[False, False])
out2 = a.copy()
out2.mask[0] = True
inside = a <= 0.01
out2[inside] = a[inside]
print('out2.mask', out2.mask, 'a.mask', a.mask, inside)
```

For 1.21.2:

```
out2.mask [[False False]] a.mask [[False False]]
```

For main:
For 1.21.2:

```
out2.mask [[ True False]] a.mask [[False False]]
```

This bisects to 25042301ae655  https://github.com/numpy/numpy/pull/19244

Note that if we replace `out2[inside] = a[inside]` by `out2[0] = a[0]` then there is no inconsistency, so it only happens when using a boolean array addressing.

**Comment by neutrinoceros** [comment]
Thank you so much @jklymak, I edited the OP with a modified version of your pure-numpy test that I personally find is a little easier to follow still.

@seberg thank you for triaging this. Though the bug is seen on the main (currently numpy 1.22.0.dev) branch, it's not crystal clear to me wether a patch would be required for numpy 1.21.3, but maybe my mental model of numpy's branch management is off.

**Comment by mattip** [comment]
@seberg why did you mark this for 1.21.3? Was #19244 backported to the 1.21 maintenance branch?

**Comment by seberg** [comment]
Sorry, no just got it wrong, thanks for the note.

**Comment by iameskild** [comment]
@neutrinoceros @mattip Apologies for the break. I believe there should been a check in place to ensure that a complete copy (such as `b[mask] = a[mask]`) works, regardless of whether the `mask` has a boolean index. 
I submitted PR #19740 for this fix.

**Comment by neutrinoceros** [comment]
I submitted a fix first in #19736, if you don't mind I was hoping to get my "contributor" badge here (I spent more than a day-worth of work tracking this down from yt).
